activejob/lib/active_job/queue_adapters/async_adapter.rb

```diff
@@ -1,22 +1,118 @@
-require 'active_job/async_job'
+require 'securerandom'
+require 'concurrent/scheduled_task'
+require 'concurrent/executor/thread_pool_executor'
+require 'concurrent/utility/processor_counter'
 
 module ActiveJob
   module QueueAdapters
     # == Active Job Async adapter
     #
-    # When enqueuing jobs with the Async adapter the job will be executed
-    # asynchronously using {AsyncJob}[http://api.rubyonrails.org/classes/ActiveJob/AsyncJob.html].
+    # The Async adapter runs jobs with an in-process thread pool.
     #
-    # To use +AsyncJob+ set the queue_adapter config to +:async+.
+    # This is the default queue adapter. It's well-suited for dev/test since
+    # it doesn't need an external infrastructure, but it's a poor fit for
+    # production since it drops pending jobs on restart.
     #
-    #   Rails.application.config.active_job.queue_adapter = :async
+    # To use this adapter, set queue adapter to +:async+:
+    #
+    #   config.active_job.queue_adapter = :async
+    #
+    # To configure the adapter's thread pool, instantiate the adapter and
+    # pass your own config:
+    #
+    #   config.active_job.queue_adapter = ActiveJob::QueueAdapters::AsyncAdapter.new \
+    #     min_threads: 1,
+    #     max_threads: 2 * Concurrent.processor_count,
+    #     idletime: 600.seconds
+    #
+    # The adapter uses a {Concurrent Ruby}[https://github.com/ruby-concurrency/concurrent-ruby] thread pool to schedule and execute
+    # jobs. Since jobs share a single thread pool, long-running jobs will block
+    # short-lived jobs. Fine for dev/test; bad for production.
     class AsyncAdapter
+      # See {Concurrent::ThreadPoolExecutor}[http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/ThreadPoolExecutor.html] for executor options.
+      def initialize(**executor_options)
+        @scheduler = Scheduler.new(**executor_options)
+      end
+
       def enqueue(job) #:nodoc:
-        ActiveJob::AsyncJob.enqueue(job.serialize, queue: job.queue_name)
+        @scheduler.enqueue JobWrapper.new(job), queue_name: job.queue_name
       end
 
       def enqueue_at(job, timestamp) #:nodoc:
-        ActiveJob::AsyncJob.enqueue_at(job.serialize, timestamp, queue: job.queue_name)
+        @scheduler.enqueue_at JobWrapper.new(job), timestamp, queue_name: job.queue_name
+      end
+
+      # Gracefully stop processing jobs. Finishes in-progress work and handles
+      # any new jobs following the executor's fallback policy (`caller_runs`).
+      # Waits for termination by default. Pass `wait: false` to continue.
+      def shutdown(wait: true)
+        @scheduler.shutdown wait: wait
+      end
```

Comment:
Ah right! It's on the adapter itself, was thinking it's documented on the internal nodoc class. Will mark this one too.